File: cpp/merge_sort.cpp
#include<bits/stdc++.h>
using namespace std;

void merge(int arr[], int l, int m, int r) {
    int nl = m - l + 1;
    int nr = r - m;
    int larr[nl], rarr[nr];
    for (int i = 0; i < nl; i++) {
        larr[i] = arr[l + i];
    }
    for (int j = 0; j < nr; j++) {
        rarr[j] = arr[m + 1 + j];
    }
    int i = 0;
    int j = 0;
    int k = l;
    while (i < nl && j < nr) {
        if (larr[i] <= rarr[j]) {
            arr[k] = larr[i];
            i++;
        }
        else {
            arr[k] = rarr[j];
            j++;
        }
        k++;
    }
    while (i < nl) {
        arr[k] = larr[i];
        i++;
        k++;
    }
    while (j < nr) {
        arr[k] = rarr[j];
        j++;
        k++;
    }
}

void MergeSort(int arr[], int l, int r) {
    int m = l + ((r - l) / 2);
    if (l < r) {
        MergeSort(arr, l, m);
        MergeSort(arr, m + 1, r);
        merge(arr, l, m, r);
    }
}

int main(int argc, char* argv[]) {
    freopen(argv[1], "r", stdin);
    freopen("mergesort.txt", "w", stdout);
    vector <int> v ;
    int s;
    while (cin>>s) {
        v.push_back(s);
    }
    int n = v.size();
    int arr[n];
    for(int i = 0; i < n; i++){
        arr[i] = v[i];
    }
    MergeSort(arr, 0, n - 1);
    for (int i = 0; i < n-1; i++) {
        cout << arr[i] << endl;
    }
    cout<<arr[n-1];
    return 0;
}

File: Python/MergeSort.py
def Merge(arr, l, m, r, Index):
	n1 = m - l + 1
	n2 = r - m
	L = [0]*n1
	R = [0]*n2
	Ls = [0]*n1
	Rs = [0]*n2

	i = 0
	while(i < n1):
		L[i] = arr[l + i]
		Ls[i] = Index[l + i] 
		i += 1

	j = 0
	while(j < n2):
		R[j] = arr[m + 1 + j]
		Rs[j] = Index[m + 1 + j]
		j += 1

	i = 0     
	j = 0     
	k = l    
 
	while(i < n1 and j < n2):
		if(L[i] <= R[j]):
			arr[k] = L[i]
			Index[k] = Ls[i]
			i += 1
		else:
			arr[k] = R[j]
			Index[k] = Rs[j]
			j += 1
		k += 1

	while(i < n1):
		arr[k] = L[i]
		Index[k] = Ls[i]
		i += 1
		k += 1

	while(j < n2):
		arr[k] = R[j]
		Index[k] = Rs[j]
		j += 1
		k += 1

def MergeSort(arr,l,r,Index):
	if(l < r):
		m = (l+(r-1))//2
		MergeSort(arr, l, m,Index)
		MergeSort(arr, m+1, r,Index)
		Merge(arr, l, m, r,Index)
	
N = int(input())
l = list(map(int, input().split()))
MergeSort(l,0,N-1,[i for i in range(N)])
print(l)

